### PR TITLE
Native OOMKills add artificial RSS samples at the memory limit

### PR DESCRIPTION
**Pull request by leekathy**
In handling native OOMKills, the existing behavior only modifies the existing `ResourceMemory` recommendations by adding an artificial memory sample based on the memory request of the OOMKilled container. 

These changes preserve that behavior and also add an adjacent modification of the `ResourceRSS` recommendations by adding an artificial RSS sample based on the memory limit of the OOMKilled container. The motivation here is that the RSS recommendation should be higher than the memory limit to avoid subsequent OOMKills. Comprehensive unit tests are added.

Upon an OOMKill, an artificial OOM sample at the memory limit is added to the RSS histogram. This histogram is a binary decaying histogram, and to differentiate these from real samples at the memory limit, OOM samples are added to the next higher bucket.

Adds `AddOomSample` to the `Histogram` interface to handle custom OOM-handling behavior for samples added to the histograms. This is only implemented for the binary decaying histogram, given that it has custom OOM handling behavior as aforementioned.

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/model/container.go [hunk 1]
```diff
@@ -197,46 +200,52 @@ func (container *ContainerState) addMemorySample(sample *ContainerUsageSample, i
...
 	return true
 }
 
 // RecordOOM adds info regarding OOM event in the model as an artificial memory sample.
-func (container *ContainerState) RecordOOM(timestamp time.Time, requestedMemory ResourceAmount) error {
+func (container *ContainerState) RecordOOM(timestamp time.Time, resource ResourceName, memory ResourceAmount) error {
 	// Discard old OOM
-	if timestamp.Before(container.MemoryWindowEnd.Add(-1 * GetAggregationsConfig().MemoryAggregationInterval)) {
+	if (resource == ResourceRSS && timestamp.Before(container.lastRSSSampleStart.Add(-1*time.Hour))) ||
+		(resource == ResourceJVMHeapCommitted && timestamp.Before(container.lastJVMHeapCommittedSampleStart.Add(-1*time.Hour))) ||
```
Comment (leekathy) [inline vertical-pod-autoscaler/pkg/recommender/model/container.go:229]: Just use timestamp, not hour

File: vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go [hunk 2]
```diff
@@ -197,9 +197,19 @@ func (a *AggregateContainerState) AddSample(sample *ContainerUsageSample) {
...
+		if sample.isOOM {
+			a.AggregateRSSPeaks.AddOomSample(BytesFromMemoryAmount(sample.Usage), 1.0, sample.MeasureStart)
+		} else {
+			a.AggregateRSSPeaks.AddSample(BytesFromMemoryAmount(sample.Usage), 1.0, sample.MeasureStart)
+		}
 	case ResourceJVMHeapCommitted:
-		a.AggregateJVMHeapCommittedPeaks.AddSample(BytesFromMemoryAmount(sample.Usage), 1.0, sample.MeasureStart)
+		// Special OOM handling for binary decaying histogram.
+		if sample.isOOM {
+			a.AggregateJVMHeapCommittedPeaks.AddOomSample(BytesFromMemoryAmount(sample.Usage), 1.0, sample.MeasureStart)
```
Comment (evansheng-db) [inline vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go:209]: iiuc we don't currently add any samples for jvm heap. What is the rationale here for only adding one for native memory and rss, but not heap?

File: vertical-pod-autoscaler/pkg/recommender/model/cluster.go [hunk 3]
```diff
@@ -246,7 +246,10 @@ func (cluster *ClusterState) RecordOOM(containerID ContainerID, timestamp time.T
 // all aggregations it matches.
 func (cluster *ClusterState) AddOrUpdateVpa(apiObject *vpa_types.VerticalPodAutoscaler, selector labels.Selector) error {
 	vpaID := VpaID{Namespace: apiObject.Namespace, VpaName: apiObject.Name}
-	annotationsMap := apiObject.Annotations
+	annotationsMap := make(vpaAnnotationsMap)
+	for key, value := range apiObject.Annotations {
+		annotationsMap[key] = value
```
Comment (evansheng-db) [inline vertical-pod-autoscaler/pkg/recommender/model/cluster.go]: OOC, why was this change required?

File: vertical-pod-autoscaler/pkg/recommender/model/container.go [hunk 4]
```diff
@@ -197,46 +200,50 @@ func (container *ContainerState) addMemorySample(sample *ContainerUsageSample, i
...
-		ScaleResource(memoryUsed, GetAggregationsConfig().OOMBumpUpRatio))
 
+	if resource == ResourceMemory {
+		// Get max of the request and the recent usage-based memory peak.
+		// Omitting oomPeak here to protect against recommendation running too high on subsequent OOMs.
+		memoryUsed := ResourceAmountMax(memory, container.memoryPeak)
+		memory = ResourceAmountMax(memoryUsed+MemoryAmountFromBytes(GetAggregationsConfig().OOMMinBumpUp),
+			ScaleResource(memoryUsed, GetAggregationsConfig().OOMBumpUpRatio))
+	}
+	// Otherwise, for RSS and JVMHeapCommitte, memory is the memory limit.
```
Comment (evansheng-db) [inline vertical-pod-autoscaler/pkg/recommender/model/container.go]: nit, `JVMHeapCommitted`

**Review by evansheng-db — COMMENTED** [review]
Thanks Kathy! Mostly lgtm, outside a couple small comments

File: vertical-pod-autoscaler/pkg/recommender/model/container.go (hunk 4 above)
Comment (leekathy) [reply vertical-pod-autoscaler/pkg/recommender/model/container.go]: thanks!

File: vertical-pod-autoscaler/pkg/recommender/model/cluster.go (hunk 3 above)
Comment (leekathy) [reply vertical-pod-autoscaler/pkg/recommender/model/cluster.go]: Thanks for the catch - it was a remnant of the old oom timestamp idea. Removed it

File: vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go (hunk 2 above)
Comment (leekathy) [reply vertical-pod-autoscaler/pkg/recommender/model/aggregate_container_state.go:209]: (discussed offline, summarizing) The `jvmheap.AddOomSample` here handles the setup to add OOM samples for JVM heap the same way as RSS, i.e. add to the next higher histogram bucket, since they both recommend the max "usage" historically. 

However, this logic wouldn't be triggered yet given that we don't know what the exact JVM OOM signal will be so nothing adds JVM OOM samples. We do have real samples for vpa-test-service - @jodzga enabled container-level for this one a while back

**Review by evansheng-db — APPROVED** [review]

File: vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go [hunk 5]
```diff
@@ -155,10 +157,12 @@ func (o *observer) OnUpdate(oldObj, newObj interface{}) {
 			if oldStatus != nil && containerStatus.RestartCount > oldStatus.RestartCount {
 				oldSpec := findSpec(containerStatus.Name, oldPod.Spec.Containers)
 				if oldSpec != nil {
-					memory := oldSpec.Resources.Requests[apiv1.ResourceMemory]
-					oomInfo := OomInfo{
+					// Artificial memory sample is created based on the memory request.
+					memoryRequest := oldSpec.Resources.Requests[apiv1.ResourceMemory]
```
Comment (jodzga) [inline vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go]: Maybe we should skip this variable renaming? In general it would be a good thing to do but considering that we will have to rebase these changes on top of upstream we should probably minimize the code we modify as much as possible.

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go [hunk 6]
```diff
@@ -176,6 +176,7 @@ func newPodClients(kubeClient kube_client.Interface, resourceEventHandler cache.
 func NewPodListerAndOOMObserver(kubeClient kube_client.Interface, namespace string) (v1lister.PodLister, oom.Observer) {
 	oomObserver := oom.NewObserver()
 	podLister := newPodClients(kubeClient, oomObserver, namespace)
+	// This is a watch on native Node MemoryPressure evictions, and increases memory recommendations for evicted pods.
```
Comment (jodzga) [inline vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:179]: Thanks for documenting!

**Review by jodzga — APPROVED** [review]
LGTM! Only minor comments.

File: vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go (hunk 5 above)
Comment (leekathy) [reply vertical-pod-autoscaler/pkg/recommender/input/oom/observer.go]: Ah I totally agree, updated to minimize modifications in naming in the rest
